components/github-api-client/src/types.rs

```diff
@@ -0,0 +1,392 @@
+    pub events_url: String,
+    pub received_events_url: String,
+    #[serde(rename = "type")]
+    pub _type: String,
+    pub site_admin: bool,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct Contents {
+    pub name: String,
+    pub path: String,
+    pub sha: String,
+    pub size: usize,
+    pub url: String,
+    pub html_url: String,
+    pub git_url: String,
+    pub download_url: String,
+    pub content: String,
+    pub encoding: String,
+}
+
+impl Contents {
+    pub fn decode(&self) -> HubResult<Vec<u8>> {
+        base64::decode(&self.content).map_err(HubError::ContentDecode)
+    }
+}
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+#[serde(default)]
+pub struct GitHubWebhookPush {
+    /// The full Git ref that was pushed. Example: "refs/heads/master"
+    #[serde(rename = "ref")]
+    pub git_ref: String,
+    /// The SHA of the most recent commit on ref before the push
+    pub before: String,
+    /// The SHA of the most recent commit on ref after the push
+    pub after: String,
+    pub created: bool,
+    pub deleted: bool,
+    pub forced: bool,
+    pub base_ref: Option<String>,
+    pub compare: String,
+    /// An array of commit objects describing the pushed commits (The array includes a maximum
+    /// of 20 commits. If necessary, you can use the Commits API to fetch additional commits.
+    /// This limit is applied to timeline events only and isn't applied to webhook deliveries)
+    pub commits: Vec<GitHubWebhookCommit>,
+    pub head_commit: GitHubWebhookCommit,
+    pub repository: PushRepository,
+    pub pusher: GitHubOwner,
+    pub sender: GitHubWebhookSender,
+    pub installation: GitHubAppInstallation,
+}
+
+impl GitHubWebhookPush {
+    pub fn changed(&self) -> Vec<&String> {
+        let mut paths = vec![];
+        for commit in self.commits.iter() {
+            paths.extend(&commit.added);
+            paths.extend(&commit.removed);
+            paths.extend(&commit.modified);
+        }
+        paths.sort();
+        paths.dedup();
+        paths
+    }
+}
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+#[serde(default)]
+pub struct GitHubWebhookCommit {
+    pub id: String,
+    pub tree_id: String,
+    /// Whether this commit is distinct from any that have been pushed before
+    pub distinct: bool,
+    /// The commit message
+    pub message: String,
+    pub timestamp: String,
+    /// Points to the commit API resource
+    pub url: String,
+    /// The git author of the commit
+    pub author: GitHubAuthor,
+    pub committer: GitHubAuthor,
+    pub added: Vec<String>,
+    pub removed: Vec<String>,
+    pub modified: Vec<String>,
+}
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+#[serde(default)]
+pub struct GitHubAuthor {
+    /// Public name of commit author
+    pub name: String,
+    /// Public email of commit author
+    pub email: String,
+    /// Display name of commit author
+    pub username: String,
+}
```

Comment:
Are the values of these just `""` if the user doesn't have them set? Should these be `Option<String>` (at least, `email` and `name`)? Does it matter much for our usecase?

(Ditto for the other types in this module that are `String` instead of `Option<String>`.)

Reply:
When we use `Option<String>` it's when the API can send `null` for a field. I should probably remove `serde(default)` from these structs which is how you'll end up with an empty string instead of a runtime error